apps/sms/services/js/moz_mobile_message/moz_mobile_message_shim.js

```diff
@@ -0,0 +1,231 @@
+  },
+
+  onDeleted: function ms_onDeleted(e) {
+    if (e.deletedThreadIds && e.deletedThreadIds.length) {
+      service.broadcast('threads-deleted', {
+        ids: e.deletedThreadIds
+      });
+    }
+  },
+
+  /* Method */  
+
+  getMessage: function ms_getMessag(id) {
+    return mozMobileMessage.getMessage(id);
+  },
+
+  retrieveMMS: function ms_retrieveMMS(id) {
+    return mozMobileMessage.retrieveMMS(id);
+  },
+
+  send: function ms_send(recipient, content, sendOpts) {
+    return mozMobileMessage.send(recipient, content, sendOpts);
+  },
+
+  sendMMS: function ms_sendMMS(dataOpts, sendOpts) {
+    return mozMobileMessage.sendMMS(dataOpts, sendOpts);
+  },
+
+  delete: function ms_delete(id) {
+    return mozMobileMessage.delete(id);
+  },
+
+  markMessageRead: function ms_markMessageRead(id, isRead, returnReq) {
+    return mozMobileMessage.markMessageRead(id, isRead, returnReq);
+  },
+
+  getSegmentInfoForText: function ms_getSegmentInfoForText(text) {
+    return mozMobileMessage.getSegmentInfoForText(text);
+  },
+
+  /* Stream */
+
+  /**
+   * Stream wrapper for getThreads API for returning threads in chunck.
+   * @param {*} stream channel for returning thread and flag.
+   */
+
+  getThreads: function ms_getThreads(stream) {
+    var cursor = null;
+    var options = {
+      data: null,
+      end: false,
+      done: false
+    }
+
+    // WORKAROUND for bug 958738. We can remove 'try\catch' block once this bug
+    // is resolved
+    try {
+      cursor = mozMobileMessage.getThreads();
+    } catch(e) {
+      console.error('Error occurred while retrieving threads: ' + e.name);
+      options.end = true;
+      options.done = true;
+      stream.write(options);
+      return;
+    }
+
+    cursor.onsuccess = function onsuccess() {
+      if (this.result) {
+        options.data = this.result;
+        stream.write(options);
+
+        this.continue();
+        return;
+      }
+
+      options.end = true;
+      options.done = true;
+      stream.write(options);
+    };
+
+    cursor.onerror = function onerror() {
+      console.error('Reading the database. Error: ' + this.error.name);
+      options.done = true;
+      stream.write(options);
+    };
+  },
+
+  /**
+   * Stream wrapper for getMessages API for returning messagess in chunck.
+   * @param {*} stream channel for returning message and flag.
+   * @param {*} options Parameter for getMessages API.
+   */
+
+  getMessages: function ms_getMgs(stream, options) {
+    /*
+    options { 
+      filter: a MobileMessageFilter or similar object
+      invert: option to invert the selection
+    }
+
+     */
+    var invert = options.invert;
+    var filter = options.filter;
+    var cursor = mozMobileMessage.getMessages(filter, !invert);
+    var opts = {
+      data: null,
+      end: false,
+      done: false   
+    }
+
+    cursor.onsuccess = function onsuccess() {
+      if (!this.done) {
+        opts.data = this.result;
```

Comment:
The original getMessages success has more complicate logic because of _stopRenderingNextStep in conversation view. I think we can simply call stream.cancel to stop the cursor(Maybe I should implement stream.cancel in shim now?).